### PR TITLE
Non-archive mode sealing

### LINKED ISSUE #176: Non-archive mode sealing

### PR DESCRIPTION
**Pull request by Zergity**
To fix #176

### CODE REVIEW

File: params/config.go [hunk 1]
```diff
@@ -362,6 +381,16 @@ func (c *ChainConfig) IsThangLongPreparationBlock(num *big.Int) bool {
 	return c.Dccs != nil && c.Dccs.ThangLongBlock != nil && c.Dccs.IsThangLongPreparationBlock(num)
 }
 
+// IsSnapshot returns whether num represents a block number exactly at the snapshot block of an epoch.
+func (c *ChainConfig) IsSnapshot(num *big.Int) bool {
+	if c.Dccs == nil || c.Dccs.ThangLongBlock == nil {
+		return false
+	}
+	depth := new(big.Int).SetUint64(CanonicalDepth)
+	depth.Add(num, depth)
```
Comment (hadv) [inline params/config.go]: nit: the final value of `depth` variable name in the line `depth.Add(num, depth)`  is a bit confusing and does not present it's meaning.

File: params/config.go [hunk 2]
```diff
@@ -362,6 +381,16 @@ func (c *ChainConfig) IsThangLongPreparationBlock(num *big.Int) bool {
 	return c.Dccs != nil && c.Dccs.ThangLongBlock != nil && c.Dccs.IsThangLongPreparationBlock(num)
 }
 
+// IsSnapshot returns whether num represents a block number exactly at the snapshot block of an epoch.
+func (c *ChainConfig) IsSnapshot(num *big.Int) bool {
```
Comment (hadv) [inline params/config.go]:
```suggestion
func (c *ChainConfig) IsSnapshotBlock(num *big.Int) bool {
```

File: params/config.go [hunk 3]
```diff
@@ -362,6 +381,16 @@ func (c *ChainConfig) IsThangLongPreparationBlock(num *big.Int) bool {
 	return c.Dccs != nil && c.Dccs.ThangLongBlock != nil && c.Dccs.IsThangLongPreparationBlock(num)
 }
 
+// IsSnapshot returns whether num represents a block number exactly at the snapshot block of an epoch.
```
Comment (hadv) [inline params/config.go]:
```suggestion
// IsSnapshotBlock returns whether num represents a block number exactly at the snapshot block of an epoch.
```

File: core/blockchain.go [hunk 4]
```diff
@@ -959,8 +959,13 @@ func (bc *BlockChain) writeBlockWithState(block *types.Block, receipts []*types.
 	}
 	triedb := bc.stateCache.TrieDB()
 
+	// Checkpoint snapshot for ThangLong consensus.
+	snapshot := bc.Config().IsSnapshot(block.Number())
+
 	// If we're running an archive node, always flush
-	if bc.cacheConfig.Disabled {
+	// Also flush for ThangLong snapshot.
```
Comment (hadv) [inline core/blockchain.go]:
```suggestion
	// Also flush for ThangLong snapshot block.
```

File: core/blockchain.go [hunk 5]
```diff
@@ -959,8 +959,13 @@ func (bc *BlockChain) writeBlockWithState(block *types.Block, receipts []*types.
 	}
 	triedb := bc.stateCache.TrieDB()
 
+	// Checkpoint snapshot for ThangLong consensus.
```
Comment (hadv) [inline core/blockchain.go]:
```suggestion
```

File: core/blockchain.go [hunk 6]
```diff
@@ -959,8 +959,13 @@ func (bc *BlockChain) writeBlockWithState(block *types.Block, receipts []*types.
...
+	snapshot := bc.Config().IsSnapshot(block.Number())
```
Comment (hadv) [inline core/blockchain.go]:
```suggestion
```

File: core/blockchain.go [hunk 7]
```diff
@@ -959,8 +959,13 @@ func (bc *BlockChain) writeBlockWithState(block *types.Block, receipts []*types.
...
+
 	// If we're running an archive node, always flush
-	if bc.cacheConfig.Disabled {
+	// Also flush for ThangLong snapshot.
+	// TODO: optimize this by gc-ing the past snapshots.
+	if bc.cacheConfig.Disabled || snapshot {
```
Comment (hadv) [inline core/blockchain.go]:
```suggestion
	if bc.cacheConfig.Disabled || bc.Config().IsSnapshotBlock(block.Number()) {
```

File: params/config.go (hunk 1 above)
Comment (Zergity) [reply params/config.go]: good point

File: core/blockchain.go (hunk 7 above)
Comment (Zergity) [reply core/blockchain.go]: I would like to keep the comment, and separate condition variable is also more clear.

Comment (hadv) [reply core/blockchain.go]: upto you, but need to change the method name to `IsSnapshotBlock()`

Comment (Zergity) [reply core/blockchain.go]: Done.

File: core/blockchain.go (hunk 6 above)
Comment (hadv) [reply core/blockchain.go]: I mean we need to update this code; otherwise build error?

**Pull request merged**